modules/gapi/src/streaming/onevpl/accelerators/dx11_alloc_resource.cpp

```diff
@@ -0,0 +1,404 @@
+
+#ifdef HAVE_ONEVPL
+#include "streaming/onevpl/accelerators/dx11_alloc_resource.hpp"
+#include "streaming/onevpl/accelerators/utils/shared_lock.hpp"
+#include "logger.hpp"
+
+#ifdef HAVE_DIRECTX
+#ifdef HAVE_D3D11
+
+namespace cv {
+namespace gapi {
+namespace wip {
+namespace onevpl {
+
+LockAdapter::LockAdapter(mfxFrameAllocator origin_allocator) :
+    lockable_allocator(origin_allocator),
+    impl() {
+    GAPI_DbgAssert((lockable_allocator.Lock && lockable_allocator.Unlock) &&
+                   "Cannot create LockAdapter for empty origin allocator");
+
+    // abandon unusable c-allocator interfaces
+    // because LockAdapter requires Lock & Unlock only
+    lockable_allocator.Alloc = nullptr;
+    lockable_allocator.Free = nullptr;
+    lockable_allocator.pthis = nullptr;
+}
+
+size_t LockAdapter::read_lock(mfxMemId mid, mfxFrameData &data) {
+    size_t prev_lock_count = 0;
+    if (impl) {
+        prev_lock_count = impl->shared_lock();
+    }
+
+    // dispatch to VPL allocator using READ access mode
+    mfxStatus sts = MFX_ERR_LOCK_MEMORY;
+    try {
+        sts = lockable_allocator.Lock(nullptr, mid, &data);
+    } catch(...) {
+    }
+
+    // adapter will throw error if VPL frame allocator fails
+    if (sts != MFX_ERR_NONE) {
+        impl->unlock_shared();
+        GAPI_Assert(false && "Cannot lock frame on READ using VPL allocator");
+    }
+
+    return prev_lock_count;
+}
+
+size_t LockAdapter::unlock_read(mfxMemId mid, mfxFrameData &data) {
+    GAPI_DbgAssert(!impl || !is_write_acquired() &&
+                   "Reject `unlock_read` in `write_lock` state");
+    lockable_allocator.Unlock(nullptr, mid, &data);
+    return impl ? impl->unlock_shared() : 0;
+}
+
+void LockAdapter::write_lock(mfxMemId mid, mfxFrameData &data) {
+    if (impl) {
+        // TODO consider using `try_lock` in loop with limited iteration count
+        // to prevent dead-lock with WARN at least notification
+        impl->lock();
+    }
+
+    // dispatch to VPL allocator using READ access mode
+    mfxStatus sts = MFX_ERR_LOCK_MEMORY;
+    try {
+        sts = lockable_allocator.Lock(nullptr, mid, &data);
+    } catch(...) {
+    }
+
+    // adapter will throw error if VPL frame allocator fails
+    if (sts != MFX_ERR_NONE) {
+        impl->unlock();
+        GAPI_Assert(false && "Cannot lock frame on WRITE using VPL allocator");
+    }
+}
+
+bool LockAdapter::is_write_acquired() {
+    if(!impl) return true;
+    return impl->owns();
+}
+
+void LockAdapter::unlock_write(mfxMemId mid, mfxFrameData &data) {
+    GAPI_DbgAssert(is_write_acquired() &&
+                   "Reject `unlock_write` for unlocked state");
+    lockable_allocator.Unlock(nullptr, mid, &data);
+    if (impl) {
+        impl->unlock();
+    }
+}
+
+SharedLock* LockAdapter::set_adaptee(SharedLock* new_impl) {
+    SharedLock* old_impl = impl;
+    impl = new_impl;
+    return old_impl;
+}
+
+SharedLock* LockAdapter::get_adaptee() {
+    return impl;
+}
```

Comment:
So if you have an `rwlock` here, can this structure be generalized or why it is tightly bound to `mfx`?

Reply:
it adaptee MediaFrameAdapter locking by R/W  to mfx callback on_lock/on_unlock in DX11AccelPolicy 

This adapter change locking behavior. when it is not set, then callback from VPL decode on_lock/on_unlock behaves  as exclusive mutex, but when DX11MediaAdapter is created then on_lock/ob_unlock would work as shared mutex